### ISSUE TITLE
Slow decrease in updateOperations over time

### ISSUE THREAD

**Issue by dcharleston** [go-graphite/go-carbon#344]
Using go-carbon 0.13.0

Starting about a week ago, all our go-carbon hosts have slowly but steadily seen their updateOperations decrease. There weren't any updates/config changes/hardware changes/etc to explain what we're seeing. The decrease started within minutes across all go-carbon hosts.

Something very odd to note, when go-carbon is restarted, it continues running at the same exact updateOperations as it was running at prior to the restart. For example, our max-updates-per-second is 650, updateOperations have fallen to 100 over the last 6 days, the host is restarted or go-carbon is restarted and updateOperations returns to 100. We have the dump file enabled. When that is disabled or we delete the dump file before starting go-carbon updateOperations resume at our max of 650 though it steadily declines still.

Is there some sort of state saved in the dump file that determines what updateOperations should be? We have plenty of disk IO available, and we can take advantage of it when restarting go-carbon without the dump file since updateOperations does in fact go back up to our limit. We have been running at our updateOperations limit for almost two years without any deviation until now.

I appreciate any feedback or thoughts! Thank you

<img width="1609" alt="Screen Shot 2020-06-24 at 11 11 23 AM" src="https://user-images.githubusercontent.com/4481337/85613047-53cd9200-b60e-11ea-9d48-c80dcfa0e807.png">

```
[common]
# Run as user. Works only in daemon mode
user = "graphite"
# Prefix for store all internal go-carbon graphs. Supported macroses: {host}
graph-prefix = "carbon.agents.graphite-central-a-gocarbon-prod-sc-08"
# Endpoint for store internal carbon metrics. Valid values: "" or "local", "tcp://host:port", "udp://host:port"
metric-endpoint = "local"
# Interval of storing internal metrics. Like CARBON_METRIC_INTERVAL
metric-interval = "1m0s"
# Increase for configuration with multi persisters
max-cpu = 8

[whisper]
enabled = true
data-dir = "/opt/graphite/storage/whisper"
# http://graphite.readthedocs.org/en/latest/config-carbon.html#storage-schemas-conf. Required
schemas-file = "/etc/go-carbon/storage-schemas.conf"
# http://graphite.readthedocs.org/en/latest/config-carbon.html#storage-aggregation-conf. Optional
aggregation-file = "/etc/go-carbon/storage-aggregations.conf"
# Workers count. Metrics sharded by "crc32(metricName) % workers"
workers = 8
# Limits the number of whisper update_many() calls per second. 0 - no limit
max-updates-per-second = 1000
# Softly limits the number of whsiper files that get created each second. 0 - no limit
max-creates-per-second = 20
# Make max-creates-per-second a hard limit. Extra new metrics are dropped. A hard throttle of 0 drops all new metrics.
hard-max-creates-per-second = false
# Sparse file creation
sparse-create = false
# use flock on every file call (ensures consistency if there are concurrent read/writes to the same file)
flock = false
# Use hashed filenames for tagged metrics instead of human readable
# https://github.com/lomik/go-carbon/pull/225
hash-filenames = true

[cache]
# Limit of in-memory stored points (not metrics)
max-size = 145000000
# Strategy to persist metrics. Values: "max","sorted","noop"
#   "max" - write metrics with most unwritten datapoints first
#   "sorted" - sort by timestamp of first unwritten datapoint.
#   "noop" - pick metrics to write in unspecified order,
#            requires least CPU and improves cache responsiveness
write-strategy = "noop"

[udp]
listen = ":2003"
enabled = true
# Enable optional logging of incomplete messages (chunked by MTU)
log-incomplete = false
# Optional internal queue between receiver and cache
buffer-size = 0

[tcp]
listen = ":2003"
enabled = true
# Optional internal queue between receiver and cache
buffer-size = 750000000

[pickle]
listen = ":2004"
enabled = true
# Limit message size for prevent memory overflow
max-message-size = 67108864
# Optional internal queue between receiver and cache
buffer-size = 0

[carbonlink]
listen = "127.0.0.1:7002"
enabled = false
# Close inactive connections after "read-timeout"
read-timeout = "30s"

# grpc api
# protocol: https://github.com/lomik/go-carbon/blob/master/helper/carbonpb/carbon.proto
# samples: https://github.com/lomik/go-carbon/tree/master/api/sample
[grpc]
listen = "127.0.0.1:7003"
enabled = false 

# http://graphite.readthedocs.io/en/latest/tags.html
[tags]
enabled = false
# TagDB url. It should support /tags/tagMultiSeries endpoint
tagdb-url = "http://127.0.0.1:8000"
tagdb-chunk-size = 32
tagdb-update-interval = 100
# Directory for send queue (based on leveldb)
local-dir = "/var/lib/graphite/tagging/"
# POST timeout
tagdb-timeout = "1s"

[carbonserver]
# Please NOTE: carbonserver is not intended to fully replace graphite-web
# It acts as a "REMOTE_STORAGE" for graphite-web or carbonzipper/carbonapi
listen = ":8080" 
# Carbonserver support is still experimental and may contain bugs
# Or be incompatible with github.com/grobian/carbonserver
enabled = true 
# Buckets to track response times
buckets = 10
# carbonserver-specific metrics will be sent as counters
# For compatibility with grobian/carbonserver
metrics-as-counters = false
# Read and Write timeouts for HTTP server
read-timeout = "60s"
write-timeout = "60s"
# Enable /render cache, it will cache the result for 1 minute
query-cache-enabled = true
# 0 for unlimited
query-cache-size-mb = 0
# Enable /metrics/find cache, it will cache the result for 5 minutes
find-cache-enabled = true
# Control trigram index
#  This index is used to speed-up /find requests
#  However, it will lead to increased memory consumption
#  Estimated memory consumption is approx. 500 bytes per each metric on disk
#  Another drawback is that it will recreate index every scan-frequency interval
#  All new/deleted metrics will still be searchable until index is recreated
trigram-index = false
# carbonserver keeps track of all available whisper files
# in memory. This determines how often it will check FS
# for new or deleted metrics.
scan-frequency = "5m0s"
# Maximum amount of globs in a single metric in index
# This value is used to speed-up /find requests with
# a lot of globs, but will lead to increased memory consumption
max-globs = 100
# Fail if amount of globs more than max-globs
fail-on-max-globs = false
# graphite-web-10-mode
# Use Graphite-web 1.0 native structs for pickle response
# This mode will break compatibility with graphite-web 0.9.x
# If false, carbonserver won't send graphite-web 1.0 specific structs
# That might degrade performance of the cluster
# But will be compatible with both graphite-web 1.0 and 0.9.x
graphite-web-10-strict-mode = true
# Allows to keep track for "last time readed" between restarts, leave empty to disable
internal-stats-dir = ""
# Calculate /render request time percentiles for the bucket, '95' means calculate 95th Percentile. To disable this feature, leave the list blank
stats-percentiles = [99, 98, 95, 75, 50]

[dump]
# Enable dump/restore function on USR2 signal
enabled = false
# Directory for store dump data. Should be writeable for carbon
path = "/opt/graphite/storage/dump"
# Restore speed. 0 - unlimited
restore-per-second = 0

[pprof]
listen = "127.0.0.1:7007"
enabled = false

# Default logger
[[logging]]
# logger name
# available loggers:
# * "" - default logger for all messages without configured special logger
# @TODO
logger = ""
# Log output: filename, "stderr", "stdout", "none", "" (same as "stderr")
file = "/opt/graphite/storage/log/go-carbon.log"
# Log level: "debug", "info", "warn", "error", "dpanic", "panic", and "fatal"
level = "warn"
# Log format: "json", "console", "mixed"
encoding = "mixed"
# Log time format: "millis", "nanos", "epoch", "iso8601"
encoding-time = "iso8601"
# Log duration format: "seconds", "nanos", "string"
encoding-duration = "seconds"
```

**Comment by deniszh** [comment]
Hi @dcharleston,

Why do you think that decreasing `updateOperations` is a bad thing? Do you have other issues?
Please note that `pointsPerUpdate` is proportionally increasing over time and `committedPoints` also looks healthy.

**Comment by dcharleston** [comment]
@deniszh My main concern is that i don't know what triggered this change and eventually there will be a limit to how much we can increase cache size. Our cache size dramatically increased (from 10-20 million to 80-100 million) and we were close to hitting our limit. I already increased the limit so we aren't currently dropping metrics but it feels like we're in a vulnerable state.

**Comment by deniszh** [comment]
@dcharleston :  Ok, then I need to extend my answer a bit. Whole idea behind Whisper format was idea that you have quite intensive write stream, and limited disk write speed - so, you need some buffer (cache) to accumulate and reorder data - then you'll be able to eventually flush everything to disk. You can read this link for motivation and design decisions - https://www.aosabook.org/en/graphite.html 
It's a same tradeoff - sacrifice RAM for speed. And according to this logic and graphs above our instance doing perfect work - after accumulating some cache it was able to drop io significantly - pointsPerUpdate are increasing, so updateOperations are decreasing, as it should be.
So, it's not required that all your metrics fit in cache. It should be vice versa - knowing your disk and RAM limits you're coming with some parameters for cache size and max_updates - then after some time Carbon cache should reach some equilibrium state between incoming flow, cache and disk flushes. No need to increase cache when it fill. Well, you can - but it will never enough, whole purpose of cache is consume as much of it as possible. Will it drop metrics on the floor when cache is full? It depends, but generally it shouldn't.
So, if you worrying about RAM you should decrease cache a bit, set proper max_updates and give it a try.

**Comment by bom-d-van** [comment]
Maybe it's because of the same issue that @avereha fixed in this mr https://github.com/lomik/go-carbon/pull/319 ?

**Comment by dcharleston** [comment]
Thanks @deniszh and @bom-d-van for looking. The bugfix in 319 looks interesting, definitely worth me looking into.

I think I didn't describe my original issue accurately so I'm sorry for that. I should have stressed how sudden of a change this was and not a normal equilibrium for us. Also, we eventually realized there was a spike in creates that correlated with the time that updateOps started to decrease. Originally, since updateOps seemed to be "locked" even through service/host restarts, it didn't seem like limiting creates or increasing CPU helped the issue. The spike in creates were from a couple services emitting high cardinality metrics, increasing our creates per minute by about 10x. After disabling the dump file and sending the high cardinality creates to the blackhole on the relay, we saw the cluster stabilize, updateOps stopped decreasing. 

I'll try to describe what we think was happening and why there was a slow decrease in performance. This assumes that create operations are much more demanding that normal update operations. We are also using the "noop" write strategy.

- go-carbon starts receiving 10x the number of creates
- since create operations require more cpu time/io wait updateOps drop
- as time goes on, the cache increases due to less updateOps/sec and the cache has a higher number of creates
- updatetOps will continue to drop as long as we are receiving enough creates and there is a backlog of creates in the cache

Along with stopping high cardinality metric emission, setting a lower limit on max-creates-per-sec would help with this. We can close this issue out but I wanted to at least provide our conclusion in case anyone else runs into this behavior.
